### PR TITLE
Add Terraform module for ACI-based Azure DevOps agents

### PR DESCRIPTION
**Pull request by jcorioland**
This PR:

- [x] move the backend configuration to the pipelines folder, because it's only used by pipeline
- [x] adds the terraform module used to deploy ACI-based DevOps agents
- [x] updates the pipeline to include the ACI-based DevOps agents module

### CODE REVIEW

**Comment by jcorioland** [comment]
@ams0 - I intentionally cancelled the pipeline, because the integration of the ACI-based DevOps agents module is not done yet, so it cannot run.

**Comment by ams0** [comment]
Run it locally and it added two shiny devops agents to the pool - win!

**Comment by ams0** [comment]
Changing to ams0/aci-agent:0.3 image reduced the run time from 18 to 3 minutes!

**Comment by ams0** [comment]
only thing I had to change, my container seems to accept slightly different variables:
```
    environment_variables = {
      VSTS_AGENT_INPUT_URL = "https://dev.azure.com/${var.devops_org_name}"
      VSTS_AGENT_INPUT_POOL = var.devops_pool_name
      VSTS_AGENT_INPUT_AUTH="pat"
      VSTS_AGENT_INPUT_TOKEN = var.devops_personal_access_token
      VSTS_AGENT_INPUT_AGENT= "aci-devops-agents-${var.suffix}-${count.index}"
    }
```

File: security.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## Security
+
+Microsoft takes the security of our software products and services seriously, which includes all source code repositories managed through our GitHub organizations, which include [Microsoft](https://github.com/Microsoft), [Azure](https://github.com/Azure), [DotNet](https://github.com/dotnet), [AspNet](https://github.com/aspnet), [Xamarin](https://github.com/xamarin), and [many more](https://opensource.microsoft.com/).
```
Comment (cesar-younes) [inline security.md:3]: This makes it sound like this is a Microsoft-hosted official repo. We may need to reconsider including this if the repo will be made public.

File: src/pipelines/infra.yaml [hunk 2]
```diff
@@ -15,6 +15,7 @@ variables:
   - name: tag
     value: "$(Build.BuildId)"
   - group: TFBackend
+  - group: ACIDevOps.Secrets
   - name: skipComponentGovernanceDetection
```
Comment (cesar-younes) [inline src/pipelines/infra.yaml:19]: This is not used anywhere. I'm guessing it was mistakenly copied over from the engagement.

File: README.md [hunk 3]
```diff
@@ -16,29 +20,24 @@ It deploys:
...
+
+### Extra:
+
+We added (under `src/scripts`):
 
+- A script to populate a Variable group from a set of key/values (TODO: use a YAML formatted file as input): `create_vg.sh`
+- A python module to delete agents from a pool. As the [official Azure DevOps CLI extension for Azure CLI](https://github.com/Azure/azure-devops-cli-extension) does not suppoer yet ([issue](https://github.com/Azure/azure-devops-cli-extension/issues/955)) we thought it would be handy to have a module to implemnt such functionality in a script (TODO: implement the module in a cli command)
 
-  Scenario:
+### Scenario:
```
Comment (cesar-younes) [inline README.md:38]: This is a repeat of the introduction at the top of the README but I would say this is better worded and is clearer so I would suggest to replace the intro with this. 
Also if it's possible to add graph, that would help add context. Personally I have context but for someone seeing this for the first time I think it would really help get the picture clear.

File: README.md [hunk 4]
```diff
@@ -16,29 +20,24 @@ It deploys:
...
+
+  The pipeline in `src/pipelines/infra.yaml` applies the content on the `main.tf` in `src/terrafomr` which calls the `common` module and the `aci-devops-agents` module in sequence; after that, a call to the terrafom template in `src/terraform/delete_access_policy` eliminates the access policy for the `current` service connection (SC) service principal (SP) effectively removing access to the secret to subsequent Azure DevOps runs. This way, even by accessing the SC' SP will make it impossible to access the secrets in Keyvault.
+
+
+### Extra:
+
+We added (under `src/scripts`):
 
+- A script to populate a Variable group from a set of key/values (TODO: use a YAML formatted file as input): `create_vg.sh`
+- A python module to delete agents from a pool. As the [official Azure DevOps CLI extension for Azure CLI](https://github.com/Azure/azure-devops-cli-extension) does not suppoer yet ([issue](https://github.com/Azure/azure-devops-cli-extension/issues/955)) we thought it would be handy to have a module to implemnt such functionality in a script (TODO: implement the module in a cli command)
```
Comment (cesar-younes) [inline README.md:36]: Spelling error in the second sentence, I'm guessing it was meant to be "does not support it yet"

File: README.md [hunk 5]
```diff
@@ -2,12 +2,16 @@
 
 [![Build Status](https://cavertes.visualstudio.com/VW_Sharing/_apis/build/status/aci-privateagents-kv-integration?branchName=master)](https://cavertes.visualstudio.com/VW_Sharing/_build/latest?definitionId=3&branchName=master)
 
+This repository shows how to build a scalable pool of Azure DevOps ACI-based agents that have private access to a secured Keyvault to retrieve secrets (in this case, an SSH private key) to operate over air-gapped resources in Azure. 
 
-This repo shows how to build a scalable pool of Azure DevOps ACI-based agents that have private access to a secured Keyvault to retrieve secrets (in this case, an SSH private key) to operate over air-gapped resources in Azure. 
+Particular attention has been placed on security: the example secrets generated and upload in the pipeline is further made unaccessible by the Azure DevOps Service Connection by locking down the Keyvault access policy.
```
Comment (cesar-younes) [inline README.md:7]: This sentence is worded in a confusing way. I suggest rewording it, here's my best attempt at that: Generating secrets and uploading them in the pipeline is not possible for the Azure DevOps Service Connection as it's blocked by the KeyVault Access Policy.

File: security.md (hunk 1 above)
Comment (jcorioland) [reply security.md:3]: @ams0 added this file, not sure why?

File: src/pipelines/infra.yaml (hunk 2 above)
Comment (jcorioland) [reply src/pipelines/infra.yaml:19]: @cesar-younes Actually it is used. All the variables related to Azure DevOps in the TF command line `-var devops_org_name=$(devopsOrganization) -var devops_personal_access_token=$(devopsPersonalAccessToken) -var devops_pool_name=$(devopsPoolName) -var devops_agent_docker_image=$(devopsAgentDockerImage) -var devops_agent_docker_tag=$(devopsAgentDockerTag)"` are coming from this group.

File: README.md (hunk 3 above)
Comment (jcorioland) [reply README.md:38]: @ams0 this is for you.

File: README.md (hunk 4 above)
Comment (jcorioland) [reply README.md:36]: @ams0 this is for you.

File: README.md (hunk 5 above)
Comment (jcorioland) [reply README.md:7]: @ams0 this is for you.

File: security.md (hunk 1 above)
Comment (ams0) [reply security.md:3]: removed it in #13

File: README.md (hunk 4 above)
Comment (ams0) [reply README.md:36]: addressed in #13

File: README.md (hunk 5 above)
Comment (ams0) [reply README.md:7]: addressed in #13

Comment (ams0) [reply README.md]: and reverted this README to original from master

**Comment by ams0** [comment]
Retested [here](https://cavertes.visualstudio.com/VW_Sharing/_build/results?buildId=147&view=logs&j=daa640cf-a2c4-58a5-7feb-9d96d0790d3d&t=daa640cf-a2c4-58a5-7feb-9d96d0790d3d) with tag=0.2-linux and working. Approving and merging.
